crates/proto/src/xfer/dns_response.rs

```diff
@@ -133,12 +132,10 @@ pub struct DnsResponse(Message);
 // TODO: when `impl Trait` lands in stable, remove this, and expose FlatMap over answers, et al.
 impl DnsResponse {
     /// Retrieves the SOA from the response. This will only exist if it was an authoritative response.
-    pub fn soa(&self) -> Option<SOA> {
+    pub fn soa(&self) -> Option<&Record> {
```

Comment:
Why is this? I can see that it's nice to avoid the clone, but it seems like returning `&SOA` would make life easier for downstream users? What do we need out of the `Record` in callers?

Reply:
I think I needed all the fields of Record. I was thinking about introducing a new `TypedRecord`... not sure if nows a good time for that.

Reply:
@djc, I started working on resolving the issue here. I think it's going to take a new type, that I'll plan to call `TypedRecord`. This will take the form of `TypedRecord<R: RecordData>` which will also introduce a new trait `RecordData` that all the `RData` types will implement. To Facilitate the reference form, I think we might also need a `TypeRecordRef<'_, R>`, which would carry a reference to the interior RecordData and Name.

So I'd like to merge this as is, and then follow up with a PR to introduce that type, which may be a bigger refactor.

Reply:
Makes sense! Is the idea that we're getting rid of the `RData` enum in favor of a trait-based (that is, open) API, or will we keep both?

Reply:
I think we'd keep both? We need something generic in the Record, and I like the known sizes of all the types. Otherwise we'd need to box the types I think... the enum is generic across them all, so I was thinking `TypedRecord<RData>` would be the generic version. You can convince me otherwise.

Reply:
Not sure if this is the best place for this discussion but would like to understand what problem `TypedRecord` and/or `TypedRecordRef` are supposed to solve.